R/variance_ratio.R

```diff
@@ -33,100 +33,169 @@
+      VR <- mean(unlist(lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var)))
+
+      ## Use cyclic_shift to calculate the null distribution
+      nullval <- cyclic_shift(df = df, time.var = time.var,
+                              species.var = species.var,
+                              abundance.var = abundance.var,
+                              replicate.var = replicate.var,
+                              FUN = variance_ratio_matrixdata,
+                              bootnumber = bootnumber)
+
+      nullout <- confint(nullval)
+
+      output <- cbind(nullout, VR)
+
     } else {
-        
-      # if multiple replicates, check all replicates have values
-        check_single(df, time.var, species.var, replicate.var)
-        df[replicate.var] <- if(is.factor(df[[replicate.var]]) == TRUE) {
-            factor(df[[replicate.var]])
-        } else {
-            df[replicate.var]
-        }
-        
-        # calculate average variance ratio across replicates
-        if(average.replicates == TRUE) {
-           check_multispp(df, species.var, replicate.var)
-            df <- df[order(df[[replicate.var]]),]
-            X <- split(df, df[replicate.var])
-            VR <- mean(unlist(lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var))) 
-        } else {
-          
-          # calculate the variance ratio for replicate
-            check_multispp(df, species.var, replicate.var)
-            df <- df[order(df[[replicate.var]]),]
-            X <- split(df, df[replicate.var])
-            VR <- do.call("rbind", lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var))
+
+      # calculate the variance ratio for replicate
+      check_multispp(df, species.var, replicate.var)
+      df <- df[order(df[[replicate.var]]),]
+      X <- split(df, df[replicate.var])
+
+      ## workaround, necessary because you are not allowed to pass an argument called FUN to lapply
+      cyclic_shift_nofun <- function(f = variance_ratio_matrixdata, ...){
+        function(...) {
+          cyclic_shift(FUN = f, ...)
         }
+      }
+
+      ## also do separate null models (ie on split list)
+      null_list <- lapply(X = X, FUN = cyclic_shift_nofun(),
+                          time.var = time.var,
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          replicate.var = NA,
+                          bootnumber = bootnumber)
+      ## calculate confidence intervals for each:
+      null_intervals <- lapply(null_list, confint)
+
+      ## combine as data.frames and preserve replicate var
+      repnames <- lapply(names(null_intervals), as.data.frame)
+      ## combine replicate names with intervals, then combine them all
+      nullout <- do.call("rbind", Map(cbind, repnames, null_intervals))
+
+      ## use the replicate.var name
+      names(nullout)[1] <- replicate.var
+
```

Comment:
These three lines of code are my replacement for `purrr::map_df`. I feel like it is a bit overcomplex, but couldn't come up with anything simpler :persevere: thoughts?